### ISSUE TITLE
await start() sometimes does not finish starting

### ISSUE THREAD

**Issue by AxelTerizaki** [j-holub/Node-MPV#79]
#### Bug Description
It's not always reproductible, but basically sometimes await start() hangs and mpv doesn't show up at all. 

##### Software Versions
 - Node-Mpv: 2.0.0-beta1
 - MPV: 0.29
 - OS: Any

##### Additional  context
I was wondering if there was a way to verbose/debug the start up process to see what is hanging exactly. For example should mpv be running / showing a window with something in side while you're waiting for all observeProperties to finish? 

If you could help in some way, I'd be grateful. For now I start mpv asynchronously, and if it doesn't start up, we have a routine that checks if it's running and starts it whenever someone uses the play/stop button.

Thanks in advance for your help :)

**Comment by j-holub** [comment]
Yeah in the start method, there are many parts where it could potentially hang. It'd be a good idea to verbose the start-up process more thoroughly.

I'm unsure about the window thing, because with the `audio-only` option, mpv doesn't show up visually at all.

By the way, 0.29 is not the latest mpv player, the latest one is 0.32.

**Comment by AxelTerizaki** [comment]
Yep, but version 0.30 and up has an issue as it uses a broken libass making \kf (karaoke fade effects) inaccurate when singing :) 

So we're sticking with 0. 29 for the time being.

Here are the options for my instance: 

```
player = new mpv(
			{
				ipc_command: '--input-ipc-server',
				auto_restart: true,
				audio_only: false,
				binary: state.binPath.mpv,
				socket: socket,
				time_update: 1,
				verbose: false,
				debug: false,
			},
			mpvOptions
		);
```

and mpvOptions is usually like this

```
let mpvOptions = [
			'--keep-open=yes',
			'--fps=60',
			'--no-border',
			'--osd-level=0',
			'--sub-codepage=UTF-8-BROKEN',
			`--log-file=${resolve(state.dataPath, 'logs/', 'mpv.log')}`,
			`--volume=${+playerState.volume}`,
			`--input-conf=${resolve(resolvedPathTemp(),'input.conf')}`,
			'--autoload-files=no'
		];
```

But as I said, quite hard to reproduce. My dev partner tried 15 times yesterday and couldn't get it to hang, and then on the 16th try made it hang for some unknown reason. :)

**Comment by AxelTerizaki** [comment]
I was going to create a new issue, but with the help of @leonekmi (who's a recent addition to the Karaoke Mugen dev team) we have more info on this issue.

He was able to pinpoint exactly what the problem is : 

Basically, when node-mpv starts mpv, it waits in stdout/stderr for the "Listen" message, and then connects and waits for an "idle" status message.

When doing mockups or small projects using node-mpv, no problem usually happens because mpv starts up quick and node-mpv can latch on the IPC connection quickly and catch the idle message.

However, with KM, which relies on multiple services besides mpv, nodeJS's single-threaded architecture can possibly slow down code execution and what happens is that node-mpv never catches the idle message, and thus never resolves the start promise, causing mpv to never show up and the whole start up sequence to hang.

This is why it's overly difficult to reproduce. On my machine which is very fast, I almost never run into the issue, but slower computers do see it occasionally.

What @leonekmi suggests to make the promise more reliable is the following patch (don't mind the first part where more info is added to debug/verbose because he needed this to find out what went wrong) : 

```
diff --git a/node_modules/node-mpv/lib/mpv/_startStop.js b/node_modules/node-mpv/lib/mpv/_startStop.js
index a02a990..3a9debd 100644
--- a/node_modules/node-mpv/lib/mpv/_startStop.js
+++ b/node_modules/node-mpv/lib/mpv/_startStop.js
@@ -109,6 +109,12 @@ const startStop = {
                                // are usually output via stdout are output via stderr instead. That's why it's required to listen
                                // for the same messages on stderr as well
                                this.mpvPlayer.stderr.on('data', stdCallback);
+
+                               if (this.options.debug || this.options.verbose) {
+                                       // Pipe mpv stdout/err to provide potential clues
+                                       this.mpvPlayer.stdout.pipe(process.stdout);
+                                       this.mpvPlayer.stderr.pipe(process.stderr);
+                               }
                        });

                        // check if mpv went into idle mode and is ready to receive commands
@@ -118,7 +124,13 @@ const startStop = {
                                // socket to check for the idle event to check if mpv fully loaded and
                                // actually running
                                const observeSocket = net.Socket();
-                               observeSocket.connect({path: this.options.socket}).on('data', (data) => {
+                               observeSocket.connect({path: this.options.socket}, () => {
+                                       // send any message to see if there's a MPV instance responding
+                                       observeSocket.write(JSON.stringify({
+                                               'command': ['get_property', 'mpv-version']
+                                       })+'\n');
+                                       if (this.options.debug || this.options.verbose) console.log('[Node-MPV] sending stimulus');
+                               }).on('data', (data) => {
                                        // parse the messages from the socket
                                        const messages = data.toString('utf-8').split('\n');
                                        // check every message
@@ -133,6 +145,14 @@ const startStop = {
                                                        //     for the rare case that somebody would pass files as input via the command line
                                                        //     through the constructor. In that case mpv never goes into idle mode
                                                        if('event' in message && ['idle','file-loaded'].includes(message.event)){
+                                                               if (this.options.debug || this.options.verbose) console.log('[Node-MPV] idling');
+                                                               observeSocket.destroy();
+                                                               resolve();
+                                                       }
+                                                       // check our stimulus response
+                                                       // Check for our stimulus with mpv-version
+                                                       if('data' in message && 'error' in message && message.error === 'success') {
+                                                               if (this.options.debug || this.options.verbose) console.log('[Node-MPV] stimulus received');
                                                                observeSocket.destroy();
                                                                resolve();
                                                        }

```

Basically, sending a harmless command like get property of mpv_version does the trick to check if the app is running, instead of relying on waiting for the "idle" message, which node-mpv can miss in some circumstances.

I think it would help to make the start promise much more reliable :)

For now we patch node-mpv during yarn install so a quick fix is not needed, but you might want to look into this with the fix provided when you have time.

**Comment by AxelTerizaki** [comment]
Also apparently `idle` got renamed to `idle-active` according to the docs https://mpv.io/manual/stable/

Not sure since when though.

**Comment by j-holub** [comment]
Thank you very much for that useful information. I will look into it. The thing I'm kinda worried about is that if the machine is too fast, it could send the message before the player is actually running, having kind of the opposite problem.

Thanks for the patch code, should help to try this out more quickly.

**Comment by AxelTerizaki** [comment]
So far I've tested this on my machine (a Ryzen 5 3600X, for what it's worth) and I've had no issues yet. At worst the code could try to retry the get-property command a few times before giving up. That's what we do with loading files for instance, using the p-retry package from sindresorhus

**Comment by blaconix** [comment]
I'm having quite a similar issue, `await mpv.start()` never gets resolved. The player opens up, but the code afterwards won't get executed. If I call the code again it hooks onto the existing mpv instance.

 Running the latest node-mpv@beta with mpv 0.32.

Edit: I can confirm that it's a mpv issue, last version that works is unofficial build `mpv-x86_64-20200426-git-640db1e`, it's broken for newer versions. Prolly the `until recently` @j-holub mentioned in the pull request comment is that exact build with the idle becoming idle-active.

**Comment by j-holub** [comment]
I have tried it locally and I'm not sure if `idle` and `idle-active` is the only problem, because I think the latest mpv I tried supported both commands. There is a real race condition happening in the start up portion of Node-MPV that @leonekmi  tried to address in #81 and I think successfully did. I haven't merged that one yet as I'm a little uncertain that everything works as it should all the time.

Have you tried to run your code with the branch from that pull request with mpv 0.32? Would be nice to know if that PR fixed the problem in your case.
